### ISSUE TITLE
Physics in a separate worker process

### ISSUE THREAD

**Issue by bobalazek** [BabylonJS/Babylon.js#6071]
Ok, I'm probably not the only thinking about that idea, but I would still like to hear the feedback/ideas/concerns of others who may have had the same thought.

I did actually got started on it, but haven't made it far yet. Still, I would like to hear the feedback first, before I spend hours and hours on it, later realising it's not possible because of xyz.

The issues I can think of right now:
* Async data (position, rotation, linear & angular velocity) -- I guess that could be simply solved by having a dummy/proxy physics impostors on the main thread, that would only update the data
* Duplicate code - it would need to require to create separate version of each PhysicsPlugin as a worker version -- not really sure if it would be possible to just simply reuse the `PhysicsEngine`  container/wrapper on the worker side? Technically it seems it would be possible with `webpack worker-loader` (https://github.com/webpack-contrib/worker-loader#examples), but not fully sure yet.

I’m sure there were/must have been more issues related to workers, as I remember babylon having a worker for the native collisions, which were discontinued I think?

**Comment by deltakosh** [comment]
This would be awesome!
The main question is how to create the worker from the current plugin code. Plugins could then be made aware that they run as a worker and then start communicating with their proxy using messages

**Comment by bobalazek** [comment]
Yes, I'll try to work on it this afternoon. Already started, but kinda got stuck on the build process, trying to include webpack worker-loader.
My idea was basically, trying to set everything in the `physicsEngine`, like for example:
```javascript
public setGravity(gravity: Vector3): void {
    this.gravity = gravity;
    this._physicsPlugin.setGravity(this.gravity);
}
// would bet
public setGravity(gravity: Vector3): void {
    this.gravity = gravity;
    if (
        this.useWorker &&
        !this.inWorkerProcess
    } {
       this.worker.postMessage('setGravity', this.gravity);
    } else {
        this._physicsPlugin.setGravity(this.gravity);
    }
}
```
and then on again have an additional function in the engine to listen to those messages and execute them (if inside the worker). Now, that sounds very very straightforward and easy, so I'm sure there must be a catch there somewhere... one thing here, that could be the issues is, that you won't be able to access the physics plugin directly anymore. For that case I think, we could also make an `AbstractPhysicsEnginePlugin` (instead of an interface) and do that part as a worker.

But somewhat I'm sure there is more to it. There was probably a reason why the worker for native collisions was removed?

Nevertheless, I'll continue this afternoon, trying to set up the worker-loader to the build process, as that was where I was kinda stuck the last time.

**Comment by deltakosh** [comment]
it was removed mostly because the communication was so slow that we missed frames. Leading to really interesting state (stuck in the wall :))

**Comment by bobalazek** [comment]
Ah, ok! :)

A question...  is there any place in the code, where you can put custom declarations?
https://github.com/webpack-contrib/worker-loader#integrating-with-typescript
```javascript
// typings/custom.d.ts
declare module "worker-loader!*" {
  class WebpackWorker extends Worker {
    constructor();
  }

  export default WebpackWorker;
}
```
Somehow I've tried everything, but can't get it to work

**Comment by deltakosh** [comment]
Let me invoke @sebavan :)

**Comment by bobalazek** [comment]
Thanks! I've included the WIP version I'm currently working on, if somebody can figure out what I'm doing wrong.

**Comment by sebavan** [comment]
All typings should go under the LibDeclarations folder as they need some special processing to ensure the d.ts we create are all ok.

**Comment by bobalazek** [comment]
Thank you!
Sadly, even though I did the changes, it's still not working. Can you maybe take a second to check if it's something obvious I'm doing wrong there, especially in webpack.config.js (https://github.com/BabylonJS/Babylon.js/pull/6077/files#diff-6d5e756cc942fd5f521d854dbb7bfe3b)?
Now, the error I get (with this configuration) is the following error:
```
Uncaught TypeError: Failed to resolve module specifier "worker-loader!./physicsWorker". Relative references must start with either "/", "./", or "../".
```
but I have added the `browser.d.ts` (https://github.com/BabylonJS/Babylon.js/pull/6077/files#diff-092c91e68c20d67996658cf89c2a81ff). I have literally tried to put it everywhere (`physicsEngine.ts`, `physicsEngineLoader.ts`, ...) but I still get the same issue

**Comment by sebavan** [comment]
Ohhh I see what you are trying to do. Unfortunately we have a couple of build target to support and webpack is not always used to build the output. For instance our es6 modules are generated directly from ts without the need to bundle or the strong dependency to one bundler on the client.

Is there any other simpler tools/scripts we could use to generate the worker ?

**Comment by bobalazek** [comment]
I only found worker-loader (https://github.com/webpack-contrib/worker-loader) and worker-plugin (https://github.com/GoogleChromeLabs/worker-plugin). Neither seems to work really.

I can't think of any other way, one could get the required modules into the worker?

**Comment by sebavan** [comment]
Could you try to create the worker outside of the default build process for now and once fully working I ll help you integrate it in the build process ?

**Comment by bobalazek** [comment]
What would be the best way to do that? I guess a separate folder/project in root folder like there is for gui, inspector, viewer, ... ?

**Comment by deltakosh** [comment]
https://playground.babylonjs.com/#N7PGG3#2

**Comment by agentdavo** [comment]
Hi!  Any further traction on phsyics in workers?  Has this found its way into 4.1? or do you know of any patches around?

**Comment by deltakosh** [comment]
Nope. Community was not too much interested by it so far

**Comment by agentdavo** [comment]
Np. Thanks for speedy response! Whilst I’m here could I possibly ask for a little advice. How would you go about an explosion and creation of an “air” pressure wave away from the explosion. Would you use force or wind or something else? 😁 How could you visualise that? Thanks!

**Comment by deltakosh** [comment]
This one is more for the forum :)

**Comment by thomlucc** [comment]
Closing for no-activity. Feel free to re-open.

**Comment by bondarenkovladislav** [comment]
Would be awesome to have calculation in worker

**Comment by syuilo** [comment]
I really want this.

Below is an unofficial Three.js example (using the wasm version of Havok), and the difference between the worker version and the non-worker version is stark.

https://lo-th.github.io/phy/index.html?E=w_havok#keva

It doesn't make the physics simulation itself smoother. However, it prevents the HTML rendering from freezing or the browser itself from hanging during complex simulations.

**Comment by sebavan** [comment]
Would you like to contribute ?

**Comment by syuilo** [comment]
Could you please reopen this issue?

**Comment by thomlucc** [comment]
@syuilo - Are you open to contribute?

**Comment by syuilo** [comment]
Unfortunately, I'm currently too busy to find the time, and I'm not familiar enough with the Babylon.js codebase to take this on. However, if I manage to implement this in my own project in the future, I'd be happy to create a pull request.

**Comment by thomlucc** [comment]
Closing as no traction. Feel free to re-open,
